### ISSUE TITLE
Spike replacement/provides constraints

### ISSUE THREAD

**Issue by jjohannes** [gradle/gradle#4208]
Once we have settled on an initial design: #4122

**Comment by melix** [comment]
So here's the result of the [spike](https://github.com/gradle/gradle/tree/cc/dependency-management/spike-replacements).

We have 3 categories of use cases:

1. _capabilities_, where the requirement is to express that a library provides the same capability as another, and therefore only one of them should be used
2. _blacklisting_, where dependency resolution should be influenced by _new information_. Typically, a vulnerability has been discovered, and a company would like to enforce the use of a fixed dependency to all consumers.
3. _variants_, where the library to be selected should be different depending on the usage

## Capabilities

Capabilities is the most important use case to cover, and corresponds to a lot of rules that can be found in the widely used [Netflix dependency rules plugin](https://github.com/nebula-plugins/gradle-resolution-rules/tree/master/src/main/resources).

The spike shows that most use cases can be solved by declaring capabilities. Again, we need to separate 2 different process:

- the library author uses Gradle to publish the library, and therefore can publish metadata that Gradle can consume.
- the library is not in control, and metadata needs to be fixed to realize that, typically, 2 libraries are mutually exclusive. This is (for now) the most common use case.

Let's start with the "fixing" use cases. We introduce the concept of `capability` in the DSL. For example, we want to express that both `cglib` and `cglib-nodep` provide the same capability, so we only want one of them on classpath:

```
    dependencies {
        constraints {
            capabilities {
                capability('cglib') {
                    providedBy 'cglib:cglib'
                    providedBy 'cglib:cglib-nodep'
                }
            }
        }
    }
```

This is a _consumer side_ declaration which declares a capability, named `cglib`, and explains that this capability is `providedBy` both the `cglib:cglib` and `cglib-nodep` modules.

If, and only if, the 2 modules are found on the dependency graph, a resolution error would occur:

> Cannot choose between cglib:cglib-nodep or cglib:cglib because they provide the same capability: cglib

Part of this work is to provide _automatic solutions_ for this. As a consequence, it should be possible to express that in that case, we prefer one version over the other:

```
    dependencies {
        constraints {
            capabilities {
                capability('cglib') {
                    prefer 'cglib:cglib'
                }
            }
        }
    }
```

This solution should _also_ work when the capabilities are actually defined in a published metadata file (it doesn't for the spike yet).

It's worth noting that we make a difference between explaining that a library "provides a capability", and the solution to that capability. While some modules _may_ want to express a preference, it should in most cases not be the case. In other words, final applications and platforms may want to publish a preference, while library authors wouldn't. This would allow, for example, a company wide preference for `slf4j-simple` over all other possible bindings, by declaring that all bindings provide the same capability.

Similarly, it is possible to solve use cases like `groovy` and `groovy-all` on classpath. See [those tests](https://github.com/gradle/gradle/blob/0b4413704c0580169ad5a1e4b9c1fcd5d9107c03/subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/CapabilitiesIntegrationTest.groovy#L205-L352) for examples.

In all cases, replacement is only performed when the 2 "conflicting" capabilities are found on the graph. If there's no conflict, the original version is used. This can be useful to realize that older libraries depend on older capabilities and that bridging libraries need to be used (like, log4j1 over log4j2).

## Blacklisting

The blacklisting use case is actually already covered by what we implemented. The idea, for example, is to force consumers to upgrade to a newer version. This can be done in 2 steps:

1. first, the consumer declares a dynamic dependency on the company wide rules platform
2. second, the platform declares constraints on dependencies

Say, for example, that we want to forcefully upgrade a dependency from 1.0.0 to 1.0.1. A typical consumer project would only have:

```
dependencies {
    implementation 'org:dep:1.0.0'
}
```
Obviously here, there's nothing that would make resolution forcefully upgrade if a newer version is published. This behavior must be enabled (1) by declaring a dependency on the platform rules:

```
dependencies {
    implementation 'org:platform-rules:latest.release'
    implementation 'org:dep:1.0.0'
}
```

Then, platform rules would declare the substitution:

```
dependencies {
   constraints {
       implementation('org:dep') {
           version {
              prefer '1.0.1'
              reject '1.0.0'
           }
           because '1.0.0 has a vulnerabililty'
       }
   }
}
```

Which would be published in Gradle metadata, hence made available to consumers. The dynamic dependency on the rules is here to make it clear that rules are live, vs static dependencies.
## Variants

It is already possible to solve those problems, as illustrated [here](https://github.com/gradle/dependency-management-samples/tree/master/experiments/replacement-constraints)

## Implementation details

The spike makes use of the existing "module replacement" API. It's not a requirement, but it makes the implementation rather straightforward. Should we consider that we can do this, or should we implement an adhoc mechanism? There are some gotchas for the implementor (likely me):

* up to now, _all_ rules were known beforehand. That is to say, replacement rules, component metadata rules, ... were all declared and known before resolution starts. This is NOT the case anymore: as we resolve new modules, new "replacements"/constraints/capabilities are discovered. This imposes some changes in the dependency graph builder, which are not necessarily obvious nor performant.
* should the capabilities be declared at the variant or the component level? The spike declares them at the variant level, but it's likely that capabilities are actually a component thing, or it involves a lot of redundant information.
* as we discover new capabilities, new constraints are added, and replacements can be triggered. However, there are some edge cases when, say, a branch of the graph is removed, and that branch contained a replacement/capability declaration. It seems at this point pretty hard to "forget" about those. This kind of enforces the recognition that capabilities are "global", and while they can be discovered as we go, they are not subject to doubt. If we say, for example, that _asm:asm_ has been relocated to _org.ow2.asm:asm_, this is true whatever the version of ASM we choose, and whatever the variant we select.

What if a library _accidently publishes wrong capabilities_?

It could be a capability that has accidental preference, or, wrong capabilities declared. In both cases, this will be solved by _rules_.

**Comment by melix** [comment]
cc @adammurdoch for veto if required

**Comment by ljacomet** [comment]
At high level I really like what you showed here and it makes sense to me.

Two remarks though:

1. If we want the normal use case to be:
* capability _declaration_ by library authors,
* capability _preference_ by application / platform authors

  Shouldn't we make the two APIs distinct?

2. > What if a library accidently publishes wrong capabilities?

    I really believe this should be "When a library publishes wrong capability?" and have a very clear story as how you fix incorrect metadata, whether they happen by accident or for more nefarious reasons.

**Comment by melix** [comment]
Declaration and preference are distinct `providedBy` vs `prefer`, but _may_ be conflated into the same platform declaration, so I feel it makes sense to have them at the same place. Doing otherwise would probably involve a larger API surface, which I'm not convinced we need/want.

**Comment by jjohannes** [comment]
Sounds good! Just to be clear, although there are three categories listed above, we now talk about implementing the first category (as the others are basically already covered). That category is **Capabilities**

My thoughts:

We will add the ability to adjust metadata in component metadata rules. That should be enough to cover the "wrong" metadata scenario. We can add that as a second step once the "capabilities" implementation is done.

I think the capability belongs on the component (NOT the variant). It _feels_ like it is making things too complex and I can't see a use case at the moment for why it should be on the variant.

It's now pretty clear (in my mind) that we should distinguish 4 things
1. Published metadata - What can/should a library author state without knowing all the possible contexts the library will be used in?
2. Component metadata rules - What should the library author have stated, but "messed up"?
3. The consuming build's dependency declaration - what do I require? (that is dependency declarations without versions or any other constraints, maybe only declarations on capabilities in the end)
4. A "platform"  which is published or defined "adhoc" as part of the build - it makes statements about other modules, i.e. introduces constraints. A dependency declaration as we know it is then a shortcut for a combination of 3. and 4.:
`implementation cglib:cglib-nodep:2.0`: _I require the _cglib_ capability. And I define, as part of my "adhoc platform", that I want the `cglib-nodep` implementation of the capability and I have the constraint that I need a version 2.0 or higher, where I prefer 2.0 if possible._

Based on this, I think _providedBy_ can be part of a library's metadata (1. or 2.) if the _capability_ is globally known. E.g. cglib and cglib-nodep are always exchangeable in every context. But I think it could also be defined in the platform 4. If the capability is valid in the context of my build/platform. And  we should be able to publish that as part of a platform definition.

I think I would do a syntax more like this:

```
dependencies {
  implementation capability('cglib') {
    //this is where I define capabilities "adhoc" as the libraries did not publish them
    providedBy 'cglib:cglib'
    providedBy 'cglib:cglib-nodep'
  }

  //alternatively, the above is published and then I depend on my platfrom
  //and i depend (require) the cglib capability
  implementation 'org:platform-rules:latest.release'
  implementation capability('cglib')

  //and then here comes my platform
  constraints {
    capability('cglib') {
      prefer 'cglib:cglib'
    }
  }

  //or the above is published, then I only need this:
  implementation 'org:platform-rules:latest.release'
}
```

The important difference to the syntax in the spike is, in my opinion, that not everything is a constraint and that you can depend directly on (require) a capability.

Looking at it this way, we then probably want the shortcut for simple builds, where you define everything directly on the dependency
```
dependencies {
  implementation capability('cglib') {
    providedBy 'cglib:cglib'
    providedBy 'cglib:cglib-nodep'
    prefer 'cglib:cglib'
  }
}
```

**Comment by melix** [comment]
I'm not fan of `implementation capability('cglib') {...}`. What does this _mean_? I can imagine that you want to say "give me a cglib, I don't care what", but it's pretty vague.

I think having:

```
dependencies {
    capabilities {
         capability('...') { ... }
    }
}
```

makes more sense. From what I understand, you worry about what we should publish, vs what we should not. So if you declare a capability to be able to choose between 2 components, do you want to publish it as part of your metadata or not? If you are declaring your _own_ capabilities, it clearly makes sense to publish. If you don't, then there are 2 options:

1. you are solving a local problem, and don't want to enforce this for your own consumers, so, you don't want to publish the capabilities
2. you are publishing a platform, in which case the constraints _are_ resolution rules

We may want to think a bit more about this...
